E2e tests for symptoms allergy and diagnosis

hrit2773 opened the pull request:
## Proposed Changes

- Fixes #11248


@ohcnetwork/care-fe-code-reviewers

## Merge Checklist

- [ ] Add specs that demonstrate bug / test a new feature.
- [ ] Update [product documentation](https://docs.ohc.network).
- [x] Ensure that UI text is kept in I18n files.
- [ ] Prep screenshot or demo video for changelog entry, and attach it to issue.
- [x] Request for Peer Reviews
- [ ] Completion of QA in Mobile Devices
- [x] Completion of QA in Desktop Devices


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Introduced end-to-end tests for managing allergies, symptoms, and diagnoses within patient encounters, including creation, editing, verification, duplicate handling, and deletion.
  * Added utility functions providing randomized clinical data for allergies, symptoms, and diagnoses to enhance testing.
  * Added support for a new "presumed" status in English localization.
  * Enhanced patient encounter workflows with new UI interactions for allergy, symptom, and diagnosis management, including adding, updating, verifying, and deleting entries.

* **Bug Fixes**
  * Improved test coverage and reliability for clinical data workflows.

* **Style**
  * Minor formatting adjustment in the allergy question component for improved readability.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

hrit2773 committed:
e2e tests for symptoms diagnosis and allergy

hrit2773 committed:
Merge branch 'e2e' of https://github.com/hrit2773/care_fe into e2e

hrit2773 committed:
Merge branch 'ohcnetwork:develop' into e2e

hrit2773 committed:
symptoms

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
Merge branch 'ohcnetwork:develop' into e2e

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
added diagnosis test

hrit2773 committed:
all tests

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 commented:
@nihal467 can you just have a look at these changes not sure why diagnosis is failing because when I tested it all the test were passing

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
updated tests

hrit2773 commented:
@nihal467 can you review this

siva222003 commented:
@hrit2773 Have you checked if the symptom or diagnosis you're trying to add already exists in the encounter? It will throw an error in such cases. Please clarify if I missed anything.

hrit2773 commented:
@siva222003 yes that's why the last and first is the difference if you see my code for allergy the position in first but for symptoms and diagnosis its last. I guess you are talking about delete? Hmm... now we cant add duplicate symptoms or diagnosis right so maybe we can select a symptom or diagnosis which does not already exist. Thanks for catching that but again not sure should I actually update that because we are selecting out of the set of few. But ya I can change that

siva222003 commented:
@hrit2773 Yeah, we need to verify whether the symptom or diagnosis exists. I was also writing tests for a questionnaire that includes these as well, but for simplicity, I'm creating a new encounter at the start and marking it as complete at the end of the test.

hrit2773 commented:
@siva222003 Hmm.. we can do that there can be many ways it will be better if Nihal confirms it or otherwise lets consider adding it but what if someone adds a symptom while the test is going on. In that case also it would break

hrit2773 commented:
otherwise we can do it anyways a new encounter gets created when you run all the tests. We can just mark the encounter as complete

siva222003 commented:
Yes, let's wait until @nihal467 shares his thoughts on this.

hrit2773 commented:
@nihal467 there has been many changes in the symptoms allergy diagnosis I need to make changes here also so marking this as draft

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
updated tests

hrit2773 committed:
lint

hrit2773 committed:
altered tags

hrit2773 committed:
scroll into view

hrit2773 committed:
change2

hrit2773 commented:
@nihal467 can you review this? I have made changes according to the latest UI changes.

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 commented:
@nihal467 the tests are not able to find ghc payyanur. Can you fix it and review this PR

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
update1

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
lint fixes"

hrit2773 committed:
updated2

hrit2773 committed:
changed2

hrit2773 committed:
refactored

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
added duplicate symptom and diagnosis verification

hrit2773 committed:
fixed login

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 commented on cypress/e2e/patient_spec/patient_encounter.cy.ts:
```diff
@@ -56,3 +65,112 @@
+      createDiagnosisDetails.diagnosisName,
+    );
+  });
+});
```
added

hrit2773 commented on cypress/e2e/patient_spec/patient_encounter.cy.ts:
```diff
@@ -56,3 +65,112 @@
+    patientEncounter.clickEditSymptoms().deleteSymptom();
+    patientPrescription.submitQuestionnaire();
+    patientEncounter.verifySymptomDelete(createSymptomsDetails.symptomName);
+  });
```
added

hrit2773 commented on cypress/e2e/patient_spec/patient_encounter.cy.ts:
```diff
@@ -56,3 +65,112 @@ describe("Patient Encounter Questionnaire", () => {
+    facilityCreation.selectFirstRandomFacility();
+  });
+
+  it("Create and edit an allergy and verify the changes", () => {
+    const createAllergyDetails = {
+      allergyName: getRandomAllergyName(),
+      criticality: "Low",
+      status: "Confirmed",
+    };
+    patientEncounter
+      .navigateToEncounters()
+      .clickInProgressEncounterFilter()
+      .openFirstEncounterDetails()
+      .clickAddAllergy()
+      .addAllergy(createAllergyDetails);
+    patientPrescription.submitQuestionnaire();
+    patientEncounter.verifyAllergy(createAllergyDetails);
+
+    const updateAllergyDetails = {
+      allergyName: createAllergyDetails.allergyName,
+      criticality: getRandomAllergyCriticality(),
+      status: getRandomAllergyStatus(),
+      notes: "Edit allergy notes",
+    };
+
+    patientEncounter.clickEditAllergy().updateAllergy(updateAllergyDetails);
+
+    patientPrescription.submitQuestionnaire();
+    patientEncounter.verifyAllergy(updateAllergyDetails);
+
+    patientEncounter.clickEditAllergy().deleteAllergy();
+    patientPrescription.submitQuestionnaire();
+    patientEncounter.verifyAllergyDelete(createAllergyDetails.allergyName);
+  });
```
anyways if we run all the tests together new encounter is being created and so the allergies will start from empty only.

hrit2773 replied:
@nihal467 your thoughts?

hrit2773 commented:
@nihal467 the page is reloading as a result of cy.find() command is it due to browser??

hrit2773 committed:
removed

hrit2773 committed:
removed some statuses

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
changed delete diagnosis logic

hrit2773 committed:
Merge branch 'e2e' of https://github.com/hrit2773/care_fe into e2e

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
updated

hrit2773 commented on cypress/pageObject/Patients/PatientEncounter.ts:
```diff
@@ -21,6 +41,247 @@ export class PatientEncounter {
+      status,
+    ]);
+
+    if (notes) {
+      cy.get('[data-cy="diagnosis-see-note"]').last().scrollIntoView();
+      cy.get('[data-cy="diagnosis-see-note"]').last().click();
+      cy.get('[data-cy="diagnosis-note"]').last().should("contain", notes);
+    }
```
because the behavior is different for symptoms and diagnosis

hrit2773 commented on cypress/pageObject/Patients/PatientEncounter.ts:
```diff
@@ -21,6 +41,247 @@
+    return this;
+  }
+
+  addSymptoms(details: SymptomDetails) {
+    const { symptomName } = details;
+    cy.get('[data-cy="add-symptom"]').scrollIntoView();
+    cy.typeAndSelectOption('[data-cy="add-symptom"]', symptomName, false);
+    return this;
+  }
+  verifyDuplicateSymptom(symptomName: string) {
+    this.addSymptoms({ symptomName });
+    cy.verifyNotification("Symptom already exists!");
+    return this;
+  }
```
duplicate allergies are allowed currently

hrit2773 commented on cypress/pageObject/Patients/PatientEncounter.ts:
```diff
@@ -21,6 +41,247 @@
+    cy.verifyNotification("Symptom already exists!");
+    return this;
+  }
+  updateSymptom(details: SymptomDetails) {
+    const { severity, status, notes } = details;
+    if (severity) {
+      cy.get('[data-cy="symptom-severity"]').last().scrollIntoView();
+      cy.clickAndSelectOption('[data-cy="symptom-severity"]', severity, {
+        position: "last",
+      });
+    }
+    if (status) {
+      cy.get('[data-cy="symptom-status"]').last().scrollIntoView();
+      cy.clickAndSelectOption('[data-cy="symptom-status"]', status, {
+        position: "last",
+      });
+    }
+    if (notes) {
+      cy.get('[data-cy="symptom-options"]').last().scrollIntoView();
+      cy.get('[data-cy="symptom-options"]').last().click();
+      cy.get('[data-cy="add-symptom-notes"]').last().click();
+      cy.typeIntoField('[data-cy="symptom-notes"]', notes, {
+        position: "last",
+        skipVerification: true,
+      });
+    }
+    return this;
+  }
```
it is correct as the behavior of symptoms and diagnosis is different

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 commented:
@nihal467 All checks are passing!! can I get a review

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
Merge branch 'develop' into e2e

nihal467 committed:
modified the common file

nihal467 reviewed (changes requested):
- add proper toaster message verification on submission
- add api verification using cy.intercept on all verification
- to avoid flakiness try, use the below approach

1. Add one symptom and verify its reflection
2. open and try to add the same (an error will be thrown, mention it exists)
3. Now, remove the existing one, then add the same symptoms again 
4. submit it and verify it presence

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
updated approach

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
fixed accidental changes

hrit2773 commented:
@nihal467 I have updated the flow of the test as suggested and also added cy.intercept() api verification. Toaster verification is handled by cy.verifyNotification in Questionnaire Submission itself. Can I get a review on this again?

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
updated delete

hrit2773 committed:
accidental

hrit2773 committed:
Merge branch 'develop' into e2e

nihal467 commented:
![image](https://github.com/user-attachments/assets/bdbb8230-2c0f-4d73-b6ec-b5ca3e1cdec9)

- when we're running the diagnosis test multiple times in local, its failing, investigate the flakiness by running it in local multiple times with fresh DB

nihal467 reviewed (changes requested):
check for flakiness

hrit2773 commented:
@nihal467 Hmm... let me check it once

hrit2773 commented:
Well... the problem is with the delete diagnosis

hrit2773 committed:
Merge branch 'develop' into e2e

AdityaJ2305 commented on cypress/utils/clinicalData.ts:
```diff
@@ -0,0 +1,86 @@
+  "Refuted",
+];
+
+export function getRandomAllergyName(): string {
+  const randomIndex = getRandomIndex(allergyNames.length);
+  return allergyNames[randomIndex];
+}
+
+export function getRandomConditionName(): string {
+  const randomIndex = getRandomIndex(conditionNames.length);
+  return conditionNames[randomIndex];
+}
+
+export function getRandomAllergyStatus(): string {
+  const randomIndex = getRandomIndex(allergyStatus.length);
+  return allergyStatus[randomIndex];
+}
+
+export function getRandomConditionStatus(): string {
+  const randomIndex = getRandomIndex(conditionStatus.length);
+  return conditionStatus[randomIndex];
+}
+
+export function getRandomAllergyCriticality(): string {
+  const randomIndex = getRandomIndex(allergyCriticality.length);
+  return allergyCriticality[randomIndex];
+}
+
+export function getRandomSymptomSeverity(): string {
+  const randomIndex = getRandomIndex(symptomSeverity.length);
+  return symptomSeverity[randomIndex];
+}
+
+export function getRandomDiagnosisVerification(): string {
+  const randomIndex = getRandomIndex(diagnosisVerification.length);
+  return diagnosisVerification[randomIndex];
+}
```
@hrit2773 , Should we have single function using switch case instead of all seperate func ? 🤔

hrit2773 replied:
@AdityaJ2305 No I feel let them be separated to increase redability

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 commented:
@nihal467 due to some latest changes I will have to update the PR to modify the logic as well. So the updated ETA is 5th June. By then I'm marking this PR as draft

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
attempt

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 commented:
@nihal467 I have made changes and ran this test 3 to 4 times continuously it did not fail. Can you check and validate it. I need a review as well

hrit2773 commented on cypress/e2e/patient_spec/patient_encounter.cy.ts:
```diff
@@ -56,3 +65,141 @@ describe("Patient Encounter Questionnaire", () => {
+    patientEncounter.clickEditAllergy().deleteAllergy();
+    patientPrescription.submitQuestionnaire();
+    patientEncounter.verifyAllergyDelete(createAllergyDetails.allergyName);
+  });
```
duplicate allergies are allowed currently so no need

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 commented:
@nihal467 can you plz review and validate it from your side as well

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 commented:
@nihal467 the tests have been running successfully 7th time continuously now.... I guess we are good to go??

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
Merge branch 'develop' into e2e

rithviknishad reviewed (changes requested):
we no longer encourage adding data-cy attributes to elements

hrit2773 commented:
@rithviknishad Ohh I see.... but should I use any other identifier?

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
removed data-cy classes

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
accommodating new changes

hrit2773 committed:
minor tweak

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
conflicts

hrit2773 committed:
changed delete

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
in progress

hrit2773 committed:
fixed

hrit2773 committed:
changed update and delete

hrit2773 committed:
updated again

hrit2773 committed:
updated using data-slot

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
scrolling issue

hrit2773 committed:
button problem

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
filter

hrit2773 committed:
Merge branch 'e2e' of https://github.com/hrit2773/care_fe into e2e

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
data-slot

hrit2773 committed:
Merge branch 'e2e' of https://github.com/hrit2773/care_fe into e2e

hrit2773 committed:
filter logic

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
updated again

hrit2773 committed:
changed problem

hrit2773 committed:
clickandselect

hrit2773 committed:
changed again1

hrit2773 committed:
changed problem1

hrit2773 committed:
changed to .then()

hrit2773 committed:
delete updated

hrit2773 committed:
delete updated again

hrit2773 committed:
refactored delete

hrit2773 committed:
command

hrit2773 committed:
refactoring

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
verify item delete

hrit2773 committed:
Merge branch 'e2e' of https://github.com/hrit2773/care_fe into e2e

hrit2773 committed:
smalll fix

hrit2773 committed:
changed to last

hrit2773 committed:
increased wait time

hrit2773 committed:
removed overview changes

hrit2773 committed:
changed the lock file back

hrit2773 committed:
changed back 1

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
updated to latest

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
small updates

hrit2773 committed:
Merge branch 'e2e' of https://github.com/hrit2773/care_fe into e2e

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
new commit

hrit2773 committed:
new commit1

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
Merge branch 'e2e' of https://github.com/hrit2773/care_fe into e2e

hrit2773 committed:
find

hrit2773 committed:
Wait for the deletion to reflect in the UI

hrit2773 committed:
Merge branch 'develop' into e2e

hrit2773 committed:
update+delete

hrit2773 commented:
@nihal467 ya still figuring these timing issues will reopen this once done
